### PR TITLE
if field type is uuid set its property to jcr:uuid

### PR DESCRIPTION
**Pull request by billsworld**
Fix for https://github.com/doctrine/phpcr-odm/issues/692

### CODE REVIEW

**Comment by dantleech** [comment]
Hmm, it is strange. I am pretty sure this has worked in the past, and after a quick look the mapping should happen [here](https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/Mapping/ClassMetadata.php#L1406). So I wonder why it isn't working in this specific case (or perhaps there has been a more general regression).

**Comment by dbu** [comment]
this should indeed not be necessary. the [basic crud test](https://github.com/doctrine/phpcr-odm/blob/master/tests/Doctrine/Tests/ODM/PHPCR/Functional/BasicCrudTest.php) works without it.

@billsworld did you encounter an error without this change? any chance you can reproduce the error with a test?

**Comment by dbu** [comment]
maybe something with Field(type="uuid") vs Uuid?

**Comment by dbu** [comment]
looking at #692 i have a suspicion: with the annotation and just type, it would work. but with xml you need to specify the name of the field and then you would also have to specify property? can you try with <field type="uuid" name="uuid" property="jcr:uuid"/> ?

File: lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php [hunk 1]
```diff
@@ -145,13 +145,17 @@ protected function getPhpcrProperty($originalAlias, $odmField)
...
+        }
+
         if (empty($fieldMeta['translated'])
             || empty($this->translator[$originalAlias])
         ) {
             return array($originalAlias, $propertyName);
         }
 
-        $propertyPath = $this->translator[$originalAlias]->getTranslatedPropertyPath($originalAlias, $fieldMeta['property'], $this->locale);
+        $propertyPath = $this->translator[$originalAlias]->getTranslatedPropertyPath($originalAlias, $propertyName, $this->locale);
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php]: this change makes sense regardless of the rest, its weird that we extract the value into a variable but then only use it once.

**Comment by dbu** [comment]
reading ClassMetadata::validateAndCompleteFieldMapping it feels like uuid should work without explicitly specifying property, though. the point dan linked to hardcodes "property" to jcr:uuid, and validateAndCompleteFieldMapping only sets property if its not already set. and yes, hardcoding to jcr:uuid comes before validating.

File: lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php [hunk 2]
```diff
@@ -145,13 +145,17 @@ protected function getPhpcrProperty($originalAlias, $odmField)
 
         $propertyName = $fieldMeta['property'];
 
+        if ($fieldMeta['type'] == 'uuid') {
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterPhpcr.php]: https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/Mapping/ClassMetadata.php#L1407 should lead to this condition never being true

**Comment by billsworld** [comment]
```field type="uuid" name="uuid" property="jcr:uuid"/>``` does the trick. There was even validation if the document is not referenceable.

**Pull request closed without merging**